default-constructed-buffers/default-constructed-buffers.md

```diff
@@ -53,18 +54,21 @@ class buffer {
 The template arguments should remain the same, so that the argument can be
 rebound to a new `buffer` instance later using the copy constructor.
 
-The `is_valid()` call would allow the programmer to query whether or not
-the buffer can be used, i.e. whether or not it was constructed with a range of
-size zero. The explicit conversion operator would call this same function but
-allow its use in `if` statements.
+The `has_storage()` call would allow the programmer to query whether or not
+the buffer can be used, i.e. whether or not it was default-constructed. The
+explicit conversion operator would call this same function but allow its use
+in `if` statements.
 
 Requesting access from a default-constructed buffer should throw an exception.
-It is not meaningful to use a zero-sized allocation on-device. Since there is
-no allocation associated with the `buffer`, `cl::sycl::buffer::set_final_data`
-and `cl::sycl::buffer::set_write_back` should behave as if the `buffer` had a
-final pointer of `nullptr` at all times. The other functions in the `buffer`
-API should behave as though the buffer were constructed with a `range` of size
-zero and otherwise behave normally.
+It is not meaningful to use a default-constructed buffer on device. Since there
```

Comment:
I think we should strengthen this further to say something to the effect of it's undefined behaviour to use a default constructed buffer (i.e. one for which `has_storage` returns `false`) to create an accessor, either within a SYCL kernel function or on the host via a host accessor.

Reply:
The sentence prior to the one on line 63 here suggests that requesting access from such a buffer should throw an exception. UB to me means many different things *could* happen, whereas an exception being thrown means that (as a library implementer) I can safely request access from any and all buffers passed to me, knowing that the user will get an exception on the other end (through the async handler, presumably). UB on the other hand means that I need to be careful or I'm leaving the program in an unknown state.

What I'm saying is this feels less helpful to the programmer, when we can instead just prescribe that an exception will communicate what went wrong in a "nice" way.

Reply:
I thought the intention of the proposal was that it be undefined, so my comment was more suggesting a strengthening of the wording to say this explicitly. But I would actually prefer that this be required to throw an exception so this is fine with me 👍